modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -0,0 +1,312 @@
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defn- ^Iterable list-tables
+  "Fetch all tables (new pages are loaded automatically by the API)."
+  ([{{:keys [project-id dataset-id]} :details, :as database}]
+   (let [bq (database->client database)]
+     (list-tables bq project-id dataset-id)))
+
+  #_([{{:keys [project-id dataset-id]} :details, :as database}, ^String page-token-or-nil]
+     ;; RAR 2020-May-11 - this gets messy, because this function is used to determine if we
+     ;; can connect to the database. If a service account JSON is being used, the project ID
+     ;; field isn't required in the admin page because it's embedded in that JSON. During the
+     ;; initial save, the database entry doesn't exist yet, so we haven't pulled the project ID
+     ;; from that JSON. In this case, we need to go look it up in the credentials.
+     ;; TODO: maybe handle this better in `driver/can-connect? :bigquery-cloud-sdk`
+     (let [bq (database->client database)]
+      (list-tables db-client project-id dataset-id page-token-or-nil)))
+
+  ([^BigQuery client, ^String project-id ^String dataset-id]
+   {:pre [client (not (str/blank? dataset-id))]}
+   (.iterateAll (.listTables client (DatasetId/of project-id dataset-id) (u/varargs BigQuery$TableListOption [])))))
+
+(defmethod driver/describe-database :bigquery-cloud-sdk
+  [_ database]
+  ;; first page through all the 50-table pages until we stop getting "next page tokens"
+  (let [tables (list-tables database) #_(loop [tables [], ^TableList table-list (list-tables database)]
+                                          (let [tables (concat tables (.getTables table-list))]
+                                            (if-let [next-page-token (.getNextPageToken table-list)]
+                                              (recur tables (list-tables database next-page-token))
+                                              tables)))]
+    ;; after that convert the results to MB format
+    {:tables (set (for [^Table table tables]
+                    {:schema nil, :name (.. table getTableId getTable)}))}))
+
+(defmethod driver/can-connect? :bigquery-cloud-sdk
+  [_ details-map]
+  ;; check whether we can connect by seeing whether we have at least one Table in the iterator
+  (let [^Iterable table-iter (list-tables {:details details-map})]
+    (.. table-iter iterator hasNext)))
+
+(s/defn get-table :- Table
+  ([{{:keys [project-id dataset-id]} :details, :as database} table-id]
+   (get-table (database->client database) project-id dataset-id table-id))
+
+  ([client :- BigQuery, project-id :- (s/maybe su/NonBlankString), dataset-id :- su/NonBlankString, table-id :- su/NonBlankString]
+   (if project-id
+     (.getTable client (TableId/of project-id dataset-id table-id) (make-array BigQuery$TableOption 0))
+     (.getTable client dataset-id table-id (make-array BigQuery$TableOption 0)))))
```

Comment:
above you used `u/varargs` for this purpose. Should we be consistent? I don't care which we choose. Also a big fan of just `(def table-options ...)` and reusing that

Reply:
Yes, that's better.  Also replaced one other `make-array` instance with `u/varargs` for consistency.